### DISCUSSION TITLE
Sobre trabalhar sozinho: organização e inovação

### DISCUSSION THREAD

**Discussion by kvnol** [frontendbr/forum#1008] [Discussão]
Creio que muitos de vocês, assim como eu, em freelas ou empresas já trabalharam sozinhos num projeto ou eram o(a) único(a) dev da empresa, certo?

Nessa situação, como vocês se organizam desde o Git até a escolha e tecnologias/frameworks e/ou inovam na escolha de uma tecnologia nova sendo que possui um prazo que talvez não permita essa ousadia?

**Comment by kvnol** [comment]
> "escolha de uma tecnologia nova sendo que possui um prazo que talvez não permita essa ousadia"

pergunto isso justamente por trabalhar como dev único nas últimas empresas que passei e muitas vezes me sentir desatualizado com o resto do mercado, por não ousar na escolha de alguma tecnologia por conta de prazo, etc.

**Comment by renanlara** [comment]
Comigo acontece algo simples: tem que desenvolver o site/tema em WordPress e pronto. É praticamente sempre o mesmo, apesar de aqui onde trabalho atualmente o cliente nunca tem contato com o WordPress. Mas a ideia é deixar mais genérico para que outros Devs futuros possam dar manutenção.

A forma como eu me organizo não tem nada demais pois tudo gira em torno de mim. Então o uso do Git é algo que nem o meu chefe sabe se eu faço ou não pois ele não sabe o que é, mas eu uso para a minha própria segurança. 

Quando eu prestava serviços para uma agência que já trabalhei e lá o trabalho era em equipe, aí a situação muda da água pro vinho, não que seja melhor ou pior, depende do ponto de vista. MAs o fluxo do uso do Git era bem maior com certeza.

Com respeito a tecnologias não tenho muito o que dizer. Só usamos o WordPress.

**Comment by luizbills** [comment]
Mas nem empresa grande ousa usar novas tecnologias para clientes, isso tem que ser feito de maneira controlada. Se você trabalha sozinho, fica até mais fácil pois você pode treinar/estudar fazendo uns projetos paralelos ou reproduzindo algum que você já fez, usando a nova tecnologia que você se interessou. Assim você ganha confiança para usa-la em produção.

Eu trabalho solo como freelancer desenvolvendo plugins para WordPress, tenho várias VM e eu para ser sincero quase não uso git. Trabalho mais é criando backups regurlamente. Geralmente só crio um repo privado quando o job já foi finalizado. Uso o git constante apenas em projetos meus.

Para organização do desenvolvimento, eu uso o trello e bastante checklists.

**Comment by safirelauene** [comment]
Olá @kvnol , eu nunca cheguei a trabalhar sozinha em uma empresa mas eu acredito que a mesma organização deve ser mantida. Eu me organizaria definindo e separando as atividades pra implementar tal funcionalidade, talvez utilizando algo como o trello, e criaria uma branch no git pra cada funcionalidade implementada. Talvez eu gaste um tempo com isso no início mas eu acho que seria mais fácil depois eu focar no desenvolvimento depois de ter tudo o que precisa ser feito definido.
Uma coisa legal que em um time de uma pessoa só se consegue fazer é usar a técnica do pomodoro, aquela onde você foca 25 minutos produzindo algo sem distrações e para 5 para descansar a fim de aumentar a produtividade. Hoje eu não consigo fazer isso no trabalho pois todos do time estão sempre interagindo um com o outro para tirar dúvidas etc e acaba um interrompendo o outro.

**Comment by felipefialho** [comment]
Fala @kvnol. Sempre trabalhei com outros devs, mas por vezes fui o único Frontender da empresa, e de fato isso pode ser um problema, dada a natureza do mundo de desenvolvimento, frontenders podem ter problemas muito especificos para resolver (performance web, acessibilidade, ux...).

Portanto as minhas soluções foram:

- Participar de fóruns / grupos de discussões e compartilhar minhas dores
- Fazer muitos projetos experimentais (que no passado resultaram num [belo Codepen](https://codepen.io/lfeh) alias)
- Iniciativas open-source me ajudavam a ter contato com outras metodologias e trabalhar em equipe
- Ir em eventos e debater sobre esses temas com outros developers

A própria criação do MeetupCSS foi uma tentativa de ter um espaço para debater CSS e melhorar como profissional... 😄 

Agora respondendo sua pergunta:

O fato de você poder escolher sozinho a tecnologia pode ser maravilhoso, afinal você vai poder usar o hype que quiser. Por outro lado, a falta de pessoas para debater e dar argumentos contra sua escolha, podem de fato trazer problemas que você, sozinho, não observou naquele momento.

O que fazia para contornar um pouco isso era falar com outros profissionais da minha confiança e ver os argumentos deles, mesmo não estando a par de todas as nuances do projeto, me ajudava a enxergar algumas coisas. Mas isso não significa que não tomei decisões ruins, elas fazem parte e servem de lição para as proximas.

**Comment by cassiohalmeida** [comment]
Vou deixar minha experiência rápida. Foi quando eu trabalhei por conta e sozinho que mais aprendi. Eu precisava ser ágil, precisava de ferramentas que fossem boas e fáceis. E o melhor, eu não tinha que pedir permissão para ninguém, eu apenas usava e aprendia. Evoluí muito nessa época. Claro, isso tudo já com toda a base teórica, então eu tentava manter o olhar crítico. Foi nessa época que, por exemplo, encontrei o Laravel (ainda na v3 na época) e eu sabia que aquilo iria ser grande, estava na cara que havia um trabalho excepcional ali. Hoje em dia eu ouço, num corredor ou outro, desenvolvedor falando: "Nossa, descobri um framework, o Laravel, vai mudar nossa vida aqui na empresa". 

Agora em empresas grandes, a história é outra. Muito software legado (forçando invlusive você a não atualizar o seu SO em casos extremos) - O cliente não quer mudar porque "tá funcionando". O cliente não tem budget para montar um time de pesquisa para migração para outro stack ou framework. É muito mais burocrático. E ao meu ver, esse cenário é o mais "tóxico" - Pode criar os grandes "seniors" que só sabem Drupal 7 (Que nem é orientado a objeto, por exemplo) - só usando um exemplo. Você sempre se vê acuado porque tudo o que você dá de ideia é sempre "vamos ver, vamos ver" - Eu me policio bastante para não me "contaminar" continuo fazendo meus projetos paralelos, continuo lendo e seguindo muita gente na comunidade e estou sempre de olho no que está vindo pro mercado, ainda mais no frontend, o qual sentimos na pelo sua velocidade.

**Comment by willianjusten** [comment]
Opa @kvnol , essa é uma pergunta bem interessante, eu vou deixar aqui um pouco da minha experiência.

Eu já trabalhei com diversos tipos de freela e muitas vezes sozinho ou tendo que coordenar o time e eu mesmo desenvolver a arquitetura a ser utilizada. Então, devido a isso, eu sempre fui a procura de coisas que pudessem facilitar o meu trabalho.

Como alguns dizem:

> A necessidade faz o sapo pular.

No início dos meus freelas eu fazia muitos sites institucionais simples, coisas de 10 páginas no máximo. Então eu acabei desenvolvendo um boilerplate que eu pudesse agilizar o processo, usando preprocessador (comecei com Less, depois Sass e por fim Stylus <3) e linguagem de template (jade/pug e o nunjucks).

Cada vez que eu ia fazendo um projeto, eu ia aprimorando a estrutura e separando também os plugins que eu ia criando. Com isso eu comecei a "ganhar tempo" nos projetos e isso me fez abrir espaço para experimentar.

Ao invés de falar para o cliente que eu faria o site dele em X dias e não mais X+10 dias, eu continuei dizendo que ia durar o mesmo de antes e aí nesse tempo extra, eu ia estudando coisas novas para os próximos clientes.

Hoje, dentro da Toptal (pelo menos no time que eu estou), apesar de ser um projeto relativamente novo, ele segue alguns padrões mais antigos e não tem tanta novidade... Em empresa realmente é mais complicado de inovar, mas é por isso que eu faço projetos pessoais por fora, assim aprendo e compartilho =D

**Comment by caiogaspar** [comment]
Trabalhei sozinho por 2 anos, era o único dev da empresa. Usava git para tudo, versionar projetos é muito importante e evita muitas cagadas (principalmente porque não tem ninguém para revisar o que você faz). 

Tentava automatizar tarefas frequentes, principalmente deploy (isso economizava muitas horas do dia e evitava dores de cabeça).  

Sempre organizava o que precisava fazer com quadros no trello, isso evitava de esquecer coisas.

Quanto a escolha de tecnologias, dependia muito do prazo que eu tinha pra fazer a tarefa e o quão seguro eu estava para usar a tecnologia naquele momento.

O que sentia ser mais difícil era se atualizar, eu acompanhava alguns devs no twitter e muitos repositórios no github, principalmente esse aqui :D. Mesmo assim sentia que não era o suficiente, a troca de experiência faz muita diferença na carreira, tanto que acabei optando por sair e trabalhar em empresas com equipes maiores.

**Comment by didiraja** [comment]
Tinha o exato mesmo problema do @kvnol, estudando em paralelo e tentando evoluir por fora do trabalho. Funcionou ano passado, mas esse ano tava batendo uma sensação de atraso, que não estava aprendendo o suficiente para evoluir como profissional.

Acabei tendo a mesma atitude do @caiogaspar, me demiti sexta pra conseguir algo numa empresa com equipe e focada em Desenvolvimento. Enquanto não acho, vou estudando e ampliando meu portfolio.
